hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -1351,14 +1353,22 @@ private void publishMapEvent(int numberOfAffectedEntries, EntryEventType eventTy
     protected <T extends Result> T executeQueryInternal(Predicate predicate, Aggregator aggregator, Projection projection,
                                                         IterationType iterationType, Target target) {
         QueryEngine queryEngine = getMapQueryEngine();
-        Predicate userPredicate = predicate;
+        final Predicate userPredicate;
 
         if (predicate instanceof PartitionPredicate) {
             PartitionPredicate partitionPredicate = (PartitionPredicate) predicate;
             Data key = toData(partitionPredicate.getPartitionKey());
             int partitionId = partitionService.getPartitionId(key);
-            userPredicate = partitionPredicate.getTarget();
-            target = createPartitionTarget(partitionId);
+            if (target.mode() == TargetMode.LOCAL_NODE && !partitionService.isPartitionOwner(partitionId)
+                    || target.mode() == TargetMode.PARTITION_OWNER && !target.partitions().contains(partitionId)
+            ) {
+                userPredicate = alwaysFalse();
```

Comment:
seems we can just return without further proceeding if userPredice is always false?

Reply:
Possibly we could, but it's an edge case, not even correctly handled before. You'd need to construct an empty `QueryResult` or an empty `AggregationResult`. Looking at QueryResult constructor, it wasn't clear to me how to construct an empty one. Would be quite complex and would need quite a lot of tests to cover the code, because it might very rarely happen otherwise. This is simple enough and gives correct result.